Unable to create a contract using pyethapp v1.2.1

corbinbs opened the pull request:
Hi,

I'm just getting started with `pyethapp` so I was working through the console guide:
https://github.com/ethereum/pyethapp/wiki/The_Console
When I got to the Creating Contracts section, https://github.com/ethereum/pyethapp/wiki/The_Console#creating-contracts , I wasn't able to get the `NameReg` contract properly created so that I could later interact with it.  After reviewing some of the `pyethereum` code, it didn't seem to be going down the `create_contract` path here https://github.com/ethereum/pyethereum/blob/develop/ethereum/processblock.py#L165  
I was still able to run `eth.find_transaction(tx)` and see a result similar to what was in the create contract section of the console guide so I thought everything had been set up properly at first. However, when I tried to interact with it, I always got `''` back when I got to the `namereg.resolve(eth.coinbase)` part.   

The adjustments in this pull request allow me to create a contract based on the example in the console guide such that I can interact with it.   However, I'm not sure if these adjustments are the best way to address it.   It looks like there's a `normalize_address` function over in https://github.com/ethereum/pyethereum/blob/develop/ethereum/utils.py#L138 that appears to already handle the blank address for contract creation.   I would have just called it, but I wasn't sure about changing the `'\0' * 20` behavior when a blank is not allowed in the pyethapp version of `normalize_address`.  I'm happy to make further refinements and/or add more test coverage for this area if it looks like I'm heading down the right path to fix it up.

Thank you,

Brian

corbinbs committed:
Allow for an empty string to be passed in to normalize_address when creating contracts

corbinbs committed:
Pass allow_blank value along to _normalize_address to support contract creation

heikoheiko commented:
We had issues with this not too long ago. 
https://github.com/ethereum/pyethapp/issues/80

If you don't mind, please check what solved it back than and what may have introduced the problem this time.

A test for this would be good.

corbinbs commented:
thanks - I'll review the history on this issue and work on a test case for it as well.

corbinbs committed:
Additional refinements for normalize_address handling in console_service.py after reviewing commit history related to https://github.com/ethereum/pyethapp/issues/80

corbinbs commented:
some notes on what I've found by going back through the commit history on pyethereum and pyethapp:

it looks like https://github.com/ethereum/pyethereum/commit/03dbe9139f5d8f34ac71880f8d268b53cc8341d5 was
updated in pyethereum to allow blank address values in that `normalize_address` function.  Then
https://github.com/ethereum/pyethapp/commit/6afb024c23f45abf4966c47ad8d6345620b60691 followed it up to make use
of these adjustments in pyethapp.  In that commit, the use of `address20` was replaced with
`ethereum.utils.normalize_address` in pyethapp/console_service.py
then https://github.com/ethereum/pyethapp/commit/cec31d101e0a8144843279f67ba608ac797ee9fb followed that
up by wrapping the `ethereum.utils.normalize_address` to return `'\0' * 20` in scenarios where blank is not allowed.

Since the `ethereum.utils.normalize_address` function doesn't allow blanks by default, I'm suspecting maybe a few references of `ethereum.utils.normalize_address` in pyethapp were not adjusted to pass in `allow_blank=True` since the wrapped version of `normalize_address` includes allow_blank=True as the default. The `address20` function that was replaced by `normalize_address` seemed to always allow for a blank address. Since `address20` was used here https://github.com/ethereum/pyethapp/commit/6afb024c23f45abf4966c47ad8d6345620b60691#diff-ecd80c17b64bd5c99f36048024c02483L115
in `eth_sendTransaction` and it was originally replaced by `normalize_address(to)` (allow_blank defaults to False),
that introduced the https://github.com/ethereum/pyethapp/issues/80 issue that popped up not too long ago.
Then https://github.com/ethereum/pyethapp/commit/49167b814ca84d2deb2edbc510e01cf7adfe45c2 came along later on
to update `eth_sendTransaction` to use the `normalize_address(to, allow_blank=True)` form.  So that got that spot fixed
up.   I've not been able to locate any current usage of `address20` so I've updated this PR to remove that function
since it doesn't seem to be used any more and it may cause more confusion down the road if it remains in pyethapp/rpc_client.py  It looks like there was one other spot https://github.com/ethereum/pyethapp/commit/6afb024c23f45abf4966c47ad8d6345620b60691#diff-ecd80c17b64bd5c99f36048024c02483L29
where `address20` was previously used but not replaced with a `normalize_address(to, allow_blank=True)`  call to preserve some of that original behavior.   The PR now adjusts that also.

I've not been able to find any usages of `z_address` from pyethapp/rpc_client.py either.
I'm wondering now if the wrapped version of `normalize_address` over in pyethapp/console_service.py includes the `'\0' * 20` part to just try to get the usages of `ethereum.utils.normalize_address` (with the allow_blank=False default) to not raise an error when it hits https://github.com/ethereum/pyethereum/commit/03dbe9139f5d8f34ac71880f8d268b53cc8341d5#diff-45243e921197fa1057f25fda22cb60a5R14
I _think_ if https://github.com/ethereum/pyethapp/commit/49167b814ca84d2deb2edbc510e01cf7adfe45c2 and the adjustment for the other `address20` usage to use the `normalize_address(to, allow_blank=True)` form had come before https://github.com/ethereum/pyethapp/commit/cec31d101e0a8144843279f67ba608ac797ee9fb then contract creation would have been ok for the rpc client and the console.

Based on this research through the commit history and some local interactive testing, 
I believe this PR now addresses the original issue along with the contract creation issue 
I encountered this week when using the console to create a contract.

I'm working on adding some test cases for this now so PR builds can exercise it...

Thanks,

Brian

corbinbs committed:
console service test coverage for sending a transaction with a contract

corbinbs commented:
Hi @heikoheiko  - I've completed the review of the issue history, made some additional adjustments and started a new set of tests for the console service.   I based the console service "transaction with contract" test case on https://github.com/ethereum/pyethapp/wiki/The_Console#creating-contracts and also tried to mimic what was currently being exercised via the current json rpc test.  
It looks like that did bump up the project's code coverage a bit.
If this seems good, I may work on a separate PR to add more to the console service test module to fully exercise the code based on the examples in the nice Console wiki page.  

Thank you,

Brian

konradkonrad commented on pyethapp/tests/test_console_service.py:
```diff
@@ -0,0 +1,146 @@
+            assert self.services.chain.chain.head.difficulty == 1
+            return self.services.chain.chain.head
+
+        def rpc_request(self, method, *args):
```
What is the purpose `rpc_request`? I can not see it being used..?

corbinbs replied:
hi @konradkonrad - I believe I probably copied over more than I should have from the pyethapp/tests/test_jsonrpc.py file.   I used that as a template to get started on the console service tests.   I'll get that cleaned up this evening and update the PR.   thank you for pointing that out

Brian

konradkonrad commented:
@corbinbs thanks for your PR! Can you clarify on my question? Other than that this looks good!

> If this seems good, I may work on a separate PR to add more to the console service test module to fully exercise the code based on the examples in the nice Console wiki page. 

That would be nice! Especially contract creation + ABI-based interaction would be really helpful to have as an integration test example (same goes for using `JSONRPCClient` btw).

corbinbs committed:
Clean up extra console service test code that was included by accident

corbinbs commented:
@konradkonrad thanks again for the feedback.
I removed `rpc_request` that was accidentally copied over to the new test module.  
I appreciate the help!

Thanks,

Brian

konradkonrad commented:
@corbinbs

>  I may work on a separate PR to add more to the console service test module to fully exercise the code based on the examples in the nice Console wiki page. 

Are you working on this? We raised several issues today, that would benefit from a run-through test. If you're already at it, please create an issue that we can assign to you. If you plan to work on it later, please review the open issues to avoid duplicated work :)

corbinbs commented:
hi @konradkonrad - check out https://github.com/ethereum/pyethapp/pull/129 to see if that's helpful at all for some of those new issues
